From daniel@caiaq.de  Fri Mar 20 19:48:52 2009
From: Daniel Mack <daniel@caiaq.de>
Date: Fri, 20 Mar 2009 19:58:57 +0100
Subject: USB: allow malformed LANGID descriptors
Cc: Daniel Mack <daniel@caiaq.de>, Greg KH <greg@kroah.com>
Message-ID: <1237575537-22910-1-git-send-email-daniel@caiaq.de>


When an USB hardware does not provide a valid LANGID, fall back to value
zero which is still a reasonable default for most devices.

Signed-off-by: Daniel Mack <daniel@caiaq.de>
Signed-off-by: Greg Kroah-Hartman <gregkh@suse.de>

---
 drivers/usb/core/message.c |    6 ++----
 1 file changed, 2 insertions(+), 4 deletions(-)

--- a/drivers/usb/core/message.c
+++ b/drivers/usb/core/message.c
@@ -804,18 +804,16 @@ int usb_string(struct usb_device *dev, i
 			dev_err(&dev->dev,
 				"string descriptor 0 read error: %d\n",
 				err);
-			goto errout;
 		} else if (err < 4) {
 			dev_err(&dev->dev, "string descriptor 0 too short\n");
-			err = -EINVAL;
-			goto errout;
 		} else {
-			dev->have_langid = 1;
 			dev->string_langid = tbuf[2] | (tbuf[3] << 8);
 			/* always use the first langid listed */
 			dev_dbg(&dev->dev, "default language 0x%04x\n",
 				dev->string_langid);
 		}
+
+		dev->have_langid = 1;
 	}
 
 	err = usb_string_sub(dev, dev->string_langid, index, tbuf);
